compiler/rustc_typeck/src/check/type_ascription.rs

```diff
@@ -0,0 +1,191 @@
+use super::FnCtxt;
+use crate::expr_use_visitor::{ConsumeMode, Delegate, ExprUseVisitor};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::def::Res;
+use rustc_hir::hir_id::HirId;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_middle::hir::place::{PlaceBase, PlaceWithHirId};
+use rustc_middle::ty;
+use rustc_span::hygiene::SyntaxContext;
+use rustc_span::symbol::Ident;
+use rustc_span::Span;
+
+// Checks for type ascriptions in lvalue contexts, i.e. in situations such as
+// (x : T) = ..., &(x : T), &mut (x : T) and (x : T).foo(...), where
+// type ascriptions can potentially lead to type unsoundness problems
+struct TypeAscriptionValidator<'a, 'tcx> {
+    fcx: &'a FnCtxt<'a, 'tcx>,
+    finder: TypeAscriptionFinder<'a>,
+}
+
+impl<'a, 'tcx> Delegate<'tcx> for TypeAscriptionValidator<'a, 'tcx> {
+    fn consume(
+        &mut self,
+        _place_with_id: &PlaceWithHirId<'tcx>,
+        _diag_expr_id: hir::HirId,
+        _mode: ConsumeMode,
+    ) {
+    }
+
+    fn borrow(
+        &mut self,
+        place_with_id: &PlaceWithHirId<'tcx>,
+        _diag_expr_id: hir::HirId,
+        bk: ty::BorrowKind,
+    ) {
+        debug!(
+            "TypeAscriptionkind::borrow(place: {:?}, borrow_kind: {:?})",
+            place_with_id.place, bk
+        );
+
+        if let PlaceBase::Rvalue(id) = place_with_id.place.base {
+            if let Some(expr) = self.fcx.tcx.hir().maybe_expr(id) {
+                debug!("expr behind place: {:?}", expr);
+                self.finder.visit_expr(expr);
```

Comment:
I am, however, wondering if we really need the finder at all. Do we have an example where it's necessary? I think I expect that if there is a type ascription being borrowed/mutated, it will **always** be the base, since it is an rvalue. I hypothesize that you can just remove this.